### ISSUE TITLE
default_scope with #from changes through association FROM clause

### ISSUE THREAD

**Issue by Lumzdas** [rails/rails#28684]
I have run in to an issue where changing a models FROM clause with a default_scope
(eg. ```default_scope { from(table_name) }```) creates unexpected behavior with has_many through associations. Namely that the model with default_scope poisons another models FROM clause with it's own. In the below test trying to access payments from a Client instance creates:
```SQL
SELECT \"payments\".* FROM invoices ...
```
though it should be:
```SQL
SELECT \"payments\".* FROM \"payments\" ...
```

### Steps to reproduce (failing test)
```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails"
  gem "arel", github: "rails/arel"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :invoices, force: true do |t|
    t.integer :client_id
  end

  create_table :payments, force: true do |t|
    t.integer :invoice_id
  end

  create_table :clients, force: true
end

class Invoice < ActiveRecord::Base
  has_many :payments
  belongs_to :client

  default_scope { from(table_name) }
end

class Payment < ActiveRecord::Base
  belongs_to :invoice
end

class Client < ActiveRecord::Base
  has_many :invoices
  has_many :payments, through: :invoices
end

class BugTest < Minitest::Test
  def test_from_clause
    client = Client.create!

    # "FROM *** INNER" - (***) should be payments in both cases
    assert_equal(
      Payment.all.joins(:invoice).to_sql.split('FROM ').last.split(' INNER').first,
      client.payments.to_sql.split('FROM ').last.split(' INNER').first
    )

    Invoice.default_scopes = nil # removing default_scope fixes the issue
    assert_equal(
      Payment.all.joins(:invoice).to_sql.split('FROM ').last.split(' INNER').first,
      client.payments.to_sql.split('FROM ').last.split(' INNER').first
    )
  end
end
```

### Expected behavior
Intermediate model with modified FROM clause should not poison other model's FROM clause

### Actual behavior
They do
This behavior seems to stem from ActiveRecord::Associations::ThroughAssociation#target_scope method, namely this line:
```ruby
scope.merge!(
  relation.except(:select, :create_with, :includes, :preload, :joins, :eager_load)
)
```
adding :from to #except argument list fixes the issue, but I don't know what else it affects

### System configuration
**Rails version**: 4.2.7.1

**Ruby version**: 2.3.1

**Comment by simonasdev** [comment]
Will this be addressed any time soon? Could definitely benefit from this fix

**Comment by cpilegard** [comment]
This is still an issue in Rails 5.2.2, but it seems to have been fixed in 6.0.0.alpha after this commit: https://github.com/rails/rails/commit/d76e3e12800414551ec42e17832227b820402007

The test above has an error, but here is a currently working and simplified version:

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :invoices, force: true do |t|
    t.integer :client_id
  end

  create_table :payments, force: true do |t|
    t.integer :invoice_id
  end

  create_table :clients, force: true
end

class Invoice < ActiveRecord::Base
  has_many :payments
  belongs_to :client

  default_scope { from(table_name) }
end

class Payment < ActiveRecord::Base
  belongs_to :invoice
end

class Client < ActiveRecord::Base
  has_many :invoices
  has_many :payments, through: :invoices
end

class BugTest < Minitest::Test
  def test_from_clause
    client = Client.create!

    assert(client.payments.from_clause.value != "invoices")
  end
end
```

**Comment by kamipo** [comment]
Backported 3f97f400ad65b7db17566fb3022f39d67cb10b5a.
